src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,538 @@
+             else
+             {
+                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.GetString(SR.UnsupportedChannelInterfaceType, typeof(TChannel))));
+             }*/
+
+            return securityChannelDispatcher;
+        }
+    }
+
+    internal abstract class ServerSecurityChannel<UChannel> where UChannel : class //, IChannel
+    {
+        private static MessageFault secureConversationCloseNotSupportedFault;
+        private string secureConversationCloseAction;
+        private SecurityListenerSettingsLifetimeManager settingsLifetimeManager;
+        private bool hasSecurityStateReference;
+        
+        protected ServerSecurityChannel(SecurityServiceDispatcher securityServiceDispatcher, UChannel innerChannel, SecurityProtocol securityProtocol, SecurityListenerSettingsLifetimeManager settingsLifetimeManager)
+        // : base(channelManager, innerChannel, securityProtocol)
+        {
+            if (settingsLifetimeManager == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("settingsLifetimeManager");
+            }
+            this.settingsLifetimeManager = settingsLifetimeManager;
+            this.SecurityProtocol = securityProtocol;
+            this.OuterChannel = (IReplyChannel)innerChannel;
+        }
+
+        internal SecurityProtocol SecurityProtocol { get; set; }
+        public IReplyChannel OuterChannel { get; private set; }
+
+        private static MessageFault GetSecureConversationCloseNotSupportedFault()
+        {
+            if (secureConversationCloseNotSupportedFault == null)
+            {
+                FaultCode faultCode = FaultCode.CreateSenderFaultCode(DotNetSecurityStrings.SecureConversationCancelNotAllowedFault, DotNetSecurityStrings.Namespace);
+                FaultReason faultReason = new FaultReason(SR.Format(SR.SecureConversationCancelNotAllowedFaultReason), System.Globalization.CultureInfo.InvariantCulture);
+                secureConversationCloseNotSupportedFault = MessageFault.CreateFault(faultCode, faultReason);
+            }
+            return secureConversationCloseNotSupportedFault;
+        }
+
+        private void ThrowIfSecureConversationCloseMessage(Message message)
+        {
+            if (message.Headers.Action == this.secureConversationCloseAction)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(new MessageSecurityException(SR.Format(SR.SecureConversationCancelNotAllowedFaultReason), null, GetSecureConversationCloseNotSupportedFault()));
+            }
+        }
+
+        [SecuritySafeCritical]
+        internal SecurityProtocolCorrelationState VerifyIncomingMessage(ref Message message, TimeSpan timeout, params SecurityProtocolCorrelationState[] correlationState)
+        {
+            if (message == null)
+            {
+                return null;
+            }
+            if (this.SecurityProtocol == null)
```

Comment:
This should be an Fx.Assert as this should only happen if there's a bug in CoreWCF. Any problems with how CoreWCF is used causing this state should be caught earlier and this code should then never get reached.